### PR TITLE
Move to tanstack router

### LINKED ISSUE #236: S15-move-to-tanstack-router
# Context
Tanstack router context might help with injecting <meta> tags for better SEO. Right now we suffer from copied meta tags and h1 tags across all pages, which is screwing up our seo rankings.

TODO: Investigate tanstack router as a solution to this. We're after something like this where we can inject router context into the <head> of our html.
Right now we're stuck rendering router pages in the body.

image

### PR DESCRIPTION
**Pull request by anthony2261**
Closes #236 <3

### CODE REVIEW

File: frontend/src/components/Connection/ConnectionSelector.tsx [hunk 1]
```diff
@@ -2,16 +2,15 @@ import ConnectionImage from "./DatabaseDialectImage";
...
 import { useCreateConversation, useGetConnections } from "@/hooks";
+import { Link, useNavigate } from "@tanstack/react-router";
 
 export const ConnectionSelector = () => {
   const navigate = useNavigate();
   const [, setConversation] = useState<IConversation | null>();
   const { data } = useGetConnections();
   const createConnection = () => {
-    navigate(Routes.NewConnection);
+    navigate({ to: "/connection/new" });
```
Comment (RamiAwar) [inline frontend/src/components/Connection/ConnectionSelector.tsx:13]: Why did you remove the constants here? Didn't get it.

File: frontend/src/components/Conversation/CodeBlock.tsx [hunk 2]
```diff
@@ -73,6 +73,8 @@ const formattedCodeOrInitial = (code: string) => {
   }
 };
 
+const chatRouteApi = getRouteApi("/_app/chat/$conversationId");
```
Comment (RamiAwar) [inline frontend/src/components/Conversation/CodeBlock.tsx:76]: Hmmm why do we need this? Nothing like useParams? This is annoying and seemingly unnecessary having to define it in two places, I don't like it.

File: frontend/src/routeTree.gen.ts [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+    LandingAboutRoute,
+    LandingBlogRoute,
+    LandingFaqRoute,
+    LandingPrivacyRoute,
+  }),
+})
+
+/* prettier-ignore-end */
+
+/* ROUTE_MANIFEST_START
```
Comment (RamiAwar) [inline frontend/src/routeTree.gen.ts:187]: ?

File: frontend/src/routes/__root.tsx [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+const rootComponent = process.env.NODE_ENV === "local" ? Main : Outlet;
+
+export const Route = createRootRouteWithContext<{
+  queryClient: QueryClient;
+}>()({
+  component: rootComponent,
+  notFoundComponent: () => (
+    <div className="flex flex-col items-center mt-8" id="error-page">
+      <h1>Oops!</h1>
```
Comment (RamiAwar) [inline frontend/src/routes/__root.tsx:13]: This is currently assigned to someone actually, make sure you close that issue so that they don't waste tiem doing this. Or re-assign to yourself and mention in a comment

File: frontend/src/routes/_app.tsx [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import { createFileRoute, redirect } from "@tanstack/react-router";
+
+export const Route = createFileRoute("/_app")({
+  beforeLoad() {
+    if (process.env.NODE_ENV !== "local") {
```
Comment (RamiAwar) [inline frontend/src/routes/_app.tsx:5]: Better use Vite env vars maybe? Not sure what the diff is.

File: frontend/src/routes/_app/chat/$conversationId.tsx [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import { createFileRoute } from "@tanstack/react-router";
+import { Conversation } from "@components/Conversation/Conversation";
+
+export const Route = createFileRoute("/_app/chat/$conversationId")({
```
Comment (RamiAwar) [inline frontend/src/routes/_app/chat/$conversationId.tsx:4]: These declarations are really ugly, do we have to do them? Why can't we define things in one place and just specify the component there? 

Taking me back to old Python where we had to add __init__.pys everywhere brainlessly for reasons we don't understand.

**Comment by RamiAwar** [comment]
Other comments: 

- Didn't see any changes relating to the landing page metadata, which was the whole point of this rewrite. Did I miss this maybe? Do we now support dynamic metadata on the landing pages?

- Are we still separating landing / app? Or is landing accessible from app? This separation is not clear to me anymore, didn't notice it when scrolling through the changes

File: frontend/src/components/Connection/ConnectionSelector.tsx (hunk 1 above)
Comment (anthony2261) [reply frontend/src/components/Connection/ConnectionSelector.tsx:13]: 1- URLs in the files created inside `src/routes` are auto-generated/auto-modified based on the relative paths of the files and their file names. For example, if I simply create a file called `hello.tsx` inside `src/routes/chat/hello`, the URL will automatically be `"/chat/hello"`, and if I try to modify the URL, we get an error

2- URL strings are now autocompleted

image

File: frontend/src/components/Conversation/CodeBlock.tsx (hunk 2 above)
Comment (anthony2261) [reply frontend/src/components/Conversation/CodeBlock.tsx]: Done

File: frontend/src/routes/_app.tsx (hunk 5 above)
Comment (anthony2261) [reply frontend/src/routes/_app.tsx:5]: Maybe in another PR? Not related to this change - we've been using `process.env.NODE_ENV` for a while

File: frontend/src/routes/_app/chat/$conversationId.tsx (hunk 6 above)
Comment (anthony2261) [reply frontend/src/routes/_app/chat/$conversationId.tsx:4]: I didn't really get your comment, do you mean it would be better if we had the components inside the route files?

**Comment by anthony2261** [comment]
I need to test that auth works. Might need to make a small addition to `beforeLoad`

File: frontend/src/routes/_app.tsx (hunk 5 above)
Comment (RamiAwar) [reply frontend/src/routes/_app.tsx:5]: Idk look into it at least, make sure there is no problematic difference. Doesn't hurt to double check this

File: frontend/src/components/Connection/ConnectionSelector.tsx (hunk 1 above)
Comment (RamiAwar) [reply frontend/src/components/Connection/ConnectionSelector.tsx:13]: Okay, I really don't like this tho. Intellisense should help and not be crucial to coding. Is there a non-file based routing option? I don't see the point of having it be file based if all the files are empty and referencing a single component. It's so much fluff and downsides with no upsides for DataLine so far.

File: frontend/src/routes/_app/chat/$conversationId.tsx (hunk 6 above)
Comment (RamiAwar) [reply frontend/src/routes/_app/chat/$conversationId.tsx:4]: Like all this is doing is: {"route": <component>}.

I don't get why we have to create N files instead of one list of length N in a single file. The old version was super readable, this is not.

Comment (anthony2261) [reply frontend/src/routes/_app/chat/$conversationId.tsx:4]: Will look into code based routing tomorrow.

File: frontend/src/components/Connection/ConnectionSelector.tsx (hunk 1 above)
Comment (anthony2261) [reply frontend/src/components/Connection/ConnectionSelector.tsx:13]: What downsides?

File: frontend/src/routes/_app/chat/$conversationId.tsx (hunk 6 above)
Comment (RamiAwar) [reply frontend/src/routes/_app/chat/$conversationId.tsx:4]: Cause they're like 5 routes in one files vs 1 route in 5 files, much more readable

File: frontend/src/components/Connection/ConnectionSelector.tsx (hunk 1 above)
Comment (RamiAwar) [reply frontend/src/components/Connection/ConnectionSelector.tsx:13]: Like not having clarity as to how routing is working cause it's so distributed across many files making it harder to find stuff, and not having constants for the strings and potentially introducing bugs if you don't have intellisense

**Review by RamiAwar — APPROVED** [review]
